utils/potential_utf/src/ustr.rs

```diff
@@ -0,0 +1,243 @@
+        // Safety: PotentialUtf8 is transparent over [u8]
+        unsafe { core::mem::transmute(other) }
+    }
+
+    /// Create a [`PotentialUtf8`] from a string slice.
+    #[inline]
+    pub const fn from_str(s: &str) -> &Self {
+        Self::from_bytes(s.as_bytes())
+    }
+
+    /// Create a [`PotentialUtf8`] from boxed bytes.
+    #[inline]
+    #[cfg(feature = "alloc")]
+    pub fn from_boxed_bytes(other: Box<[u8]>) -> Box<Self> {
+        // Safety: PotentialUtf8 is transparent over [u8]
+        unsafe { core::mem::transmute(other) }
+    }
+
+    /// Create a [`PotentialUtf8`] from a boxed `str`.
+    #[inline]
+    #[cfg(feature = "alloc")]
+    pub fn from_boxed_str(other: Box<str>) -> Box<Self> {
+        Self::from_boxed_bytes(other.into_boxed_bytes())
+    }
+
+    /// Get the bytes from a [`PotentialUtf8].
+    #[inline]
+    pub const fn as_bytes(&self) -> &[u8] {
+        &self.0
+    }
+
+    /// Attempt to convert a [`PotentialUtf8`] to a `str`.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use potential_utf::PotentialUtf8;
+    ///
+    /// static A: &PotentialUtf8 = PotentialUtf8::from_bytes(b"abc");
+    ///
+    /// let b = A.try_as_str().unwrap();
+    /// assert_eq!(b, "abc");
+    /// ```
+    // Note: this is const starting in 1.63
+    #[inline]
+    pub fn try_as_str(&self) -> Result<&str, core::str::Utf8Error> {
+        core::str::from_utf8(&self.0)
+    }
+}
+
+impl<'a> From<&'a str> for &'a PotentialUtf8 {
+    #[inline]
+    fn from(other: &'a str) -> Self {
+        PotentialUtf8::from_str(other)
+    }
+}
+
+#[cfg(feature = "alloc")]
+impl From<Box<str>> for Box<PotentialUtf8> {
+    #[inline]
+    fn from(other: Box<str>) -> Self {
+        PotentialUtf8::from_boxed_str(other)
+    }
+}
+
+impl Deref for PotentialUtf8 {
+    type Target = [u8];
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+#[cfg(feature = "zerovec")]
+impl<'a> zerovec::maps::ZeroMapKV<'a> for PotentialUtf8 {
+    type Container = zerovec::VarZeroVec<'a, PotentialUtf8>;
+    type Slice = zerovec::VarZeroSlice<PotentialUtf8>;
+    type GetType = PotentialUtf8;
+    type OwnedType = Box<PotentialUtf8>;
+}
+
+// Safety (based on the safety checklist on the VarULE trait):
+//  1. PotentialUtf8 does not include any uninitialized or padding bytes (transparent over a ULE)
+//  2. PotentialUtf8 is aligned to 1 byte (transparent over a ULE)
+//  3. The impl of `validate_byte_slice()` returns an error if any byte is not valid (impossible)
+//  4. The impl of `validate_byte_slice()` returns an error if the slice cannot be used in its entirety (impossible)
+//  5. The impl of `from_byte_slice_unchecked()` returns a reference to the same data (returns the argument directly)
+//  6. All other methods are defaulted
+//  7. `[T]` byte equality is semantic equality (transparent over a ULE)
+#[cfg(feature = "zerovec")]
+unsafe impl zerovec::ule::VarULE for PotentialUtf8 {
```

Comment:
suggestion: I _think_ `derive(VarULE)` ought to work here? but it's also fine to avoid the custom derive